plugins/poll/lib/PollActions.php

```diff
@@ -0,0 +1,256 @@
+			'SINGLE_ANONYMOUS',
+			'SINGLE_RESTRICT',
+			'MULTI_ANONYMOUS',
+			'MULTI_RESTRICT');
+	}
+
+	/* Poll Id Action */
+	public static function generatePollId($type = 'SINGLE_ANONYMOUS')
+	{
+		self::init();
+		if (!self::isValidPollType($type))
+			throw new Exception("Poll type provided is invalid");
+		$randKey = rand();
+		$hash = hash_hmac('SHA256', self::$kalturaSecret, $randKey);
+		return $type.self::ID_SEPARATOR_CHAR.$hash.self::ID_SEPARATOR_CHAR.$randKey;
+	}
+
+	private static function  isValidPollType($type)
+	{
+		return in_array($type, self::$pollTypes);
+	}
+
+	private static function isValidPollIdStructure($id)
+	{
+		self::init();
+		$idElements = explode(self::ID_SEPARATOR_CHAR, $id);
+		if (count($idElements) === self::ID_NUM_ELEMENTS ) {
+			$pollType = $idElements[0];
+			$hash = $idElements[1];
+			$key = $idElements[2];
+			$simulatedHash = hash_hmac('SHA256', self::$kalturaSecret, $key);
+			$isHashOk = strcmp($hash, $simulatedHash) === 0;
+			$validPollType = self::isValidPollType($pollType);
+			return $isHashOk && $validPollType;
+
+		}
+		return false;
+	}
+
+	/* Poll Vote Actions */
+	public static function vote($params)
+	{
+		$pollId = $params['pollId'];
+		$userId = $params['userId'];
+		$ansIds = $params['answerIds'];
+		return self::setVote($pollId, $userId, $ansIds);
+	}
+
+	public static function setVote($pollId, $userId, $ansIds)
+	{
+		if (!$pollId || !$userId || !$ansIds)
+			throw new Exception('Missing parameter for vote action');
+		$answers = explode(self::ANSWER_SEPARATOR_CHAR, $ansIds);
+		if (self::isValidPollIdStructure($pollId)) {
+			// check early user vote
+			$previousAnswers = self::$pollsCacheHandler->setCacheVote($userId, $pollId, $answers);
+			if ($previousAnswers) {
+				self::$pollsCacheHandler->decrementAnsCounter($pollId, $previousAnswers);
+			} else {
+				self::$pollsCacheHandler->incrementPollVotersCount($pollId);
+			}
+			self::$pollsCacheHandler->increaseAnsCounter($pollId, $answers);
+			return "Successfully voted";
+		}
+		return "Failed to vote due to bad poll id structure";
+	}
+
+	/* Poll Get Votes Actions */
+
+	public static function getVotes($pollId, $ansIds)
+	{
+		if (!$pollId || !$ansIds)
+			throw new Exception('Missing parameter for getVotes action');
+		self::init();
+		$answers = explode(self::ANSWER_SEPARATOR_CHAR, $ansIds);
+		$pollVotes = new PollVotes($pollId);
+		$pollVotes->setNumVoters(self::$pollsCacheHandler->getPollVotersCount($pollId));
+		foreach($answers as $ansId) {
+			$answerCount = self::$pollsCacheHandler->getAnswerCounter($pollId, $ansId);
+			$pollVotes->addAnswerCounter($ansId, $answerCount);
+		}
+		return $pollVotes;
+	}
+
+}
+
+class PollCacheHandler
+{
+
+	private $cache;
+
+	private $cacheTTL;
+
+	public function __construct($cacheTTL)
+	{
+		$this->cache = kCacheManager::getSingleLayerCache(kCacheManager::CACHE_TYPE_CACHE_ONLY_ACTIONS);
+		$this->cacheTTL = $cacheTTL;
+	}
+
+	public function setCacheVote($userId, $pollId, $ansIds)

```

Comment:
This function lays in a class named PollCacheHandler - atomic is for surely not correct since there is a get and set within it - I can change the name to setVote